app/lib/pdf/finance/invoice_report_data.rb

```diff
@@ -129,4 +130,20 @@ def total_invoice_label
       [total_without_vat, total_vat, total]
     end
   end
+
+  # Depending on the storage option for the attachment:
+  # - retrieve either the full path of the local file, or the URL of the file in S3
+  # - read the file differently
+  def logo_file
+    case storage = @logo.options[:storage].to_sym
+    when :filesystem
+      # read as binary file 'b'
+      File.open(@logo.path(LOGO_ATTACHMENT_STYLE), 'rb')
+    when :s3
+      URI.open(@logo.url(LOGO_ATTACHMENT_STYLE))
+    else
+      raise StandardError, "Invalid attachment type #{storage}"
+    end
+  end
+
```

Comment:
This would need some testing, I think.

Also, how was this other test passing so far? 

[invoice_generator_test.rb#L30](https://github.com/3scale/porta/blob/7bcbd50b1cd8dc65b5ff8974142b437267238fb0/test/unit/pdf/finance/invoice_generator_test.rb#L30)

That test probably needs checking out to ensure it's working fine.

Reply:
Good catch @jlledom!
[invoice_generator_test.rb#L30](https://github.com/3scale/porta/blob/7bcbd50b1cd8dc65b5ff8974142b437267238fb0/test/unit/pdf/finance/invoice_generator_test.rb#L30) must be updated indeed. Well, the assertion is really not very informative :sweat_smile: and the methods are stubbed (which will need to be updated)

Reply:
`:filesystem` and `:s3` options are tested, I could not test the invalid attachment type. It's a bit tricky, as we rely on `paperclip` library, and I have not found a way to manipulate the attachment type (without losing the rest of the functionality)